src/org/sosy_lab/java_smt/solvers/z3/Z3FormulaCreator.java

```diff
@@ -928,12 +966,6 @@ private long goalToAST(long z3context, long goal) {
   /** Closing the context. */
   public void forceClose() {
     cleanupReferences();
-
-    // Force clean all ASTs, even those which were not GC'd yet.
-    // Is a no-op if phantom reference handling is not enabled.
-    for (long ast : referenceMap.values()) {
-      Native.decRef(getEnv(), ast);
-    }
```

Comment:
What is the reason for deleting this?

Reply:
This was by accident because I misinterpreted the purpose. I added it back.

However, I noticed that if the `PhantomReferences` are already enqueued, then they will never get removed from the queue after closing (until the context object itself gets collected).
Unfortunately, `ReferenceQueue.clear` does not exist and one has to poll all entries to clear them.
Should I add this or does this not really matter?

Reply:
You mean those references that have been enqueued before `close()` is called? These could in principle be cleaned up, but note that there will likely also be references that are enqueued after `close()` is called, and there is nothing we can do about them, so I do not think we need to bother with the already enqueued ones.

Once the solver context becomes unreachable and is cleaned up by GC, the queue and all `PhantomReference`s are also cleaned up. So this affects only users who close the context but then keep a reference to it for a longer time.

Reply:
See my answer to the bulk deletion point you raised further below.